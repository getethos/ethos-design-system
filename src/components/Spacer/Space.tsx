import * as React from 'react'
import useInvalid from '../../hooks/useInvalid.js'
type SpaceProps = {
  height?: any
  width?: any
}
const HEIGHTS = {
  H80: 80,
  H72: 72,
  H64: 64,
  H56: 56,
  H48: 48,
  H40: 40,
  H32: 32,
  H24: 24,
  H16: 16,
  H8: 8,
  H4: 4,
  AUTO: 'auto',
}
const WIDTHS = {
  W80: 80,
  W72: 72,
  W64: 64,
  W56: 56,
  W48: 48,
  W40: 40,
  W32: 32,
  W24: 24,
  W16: 16,
  W8: 8,
  W4: 4,
  AUTO: 'auto',
}
export const Space: React.FC<SpaceProps> & {
  PUBLIC_PROPS: any
  HEIGHTS: typeof HEIGHTS
  WIDTHS: typeof WIDTHS
} = ({ height, width, ...rest }) => {
  // Verify that height and/or width were valid enum values
  const validHeight = Object.values(Space.HEIGHTS).find((h) => h === height)
  if (height && !validHeight) throw new TypeError('Invalid height.')
  const validWidth = Object.values(Space.WIDTHS).find((w) => w === width)
  if (width && !validWidth) throw new TypeError('Invalid width.')
  if (!width && !height) throw new TypeError('Must specify width or height.')
  // Verify that no invalid props were supplied
  const [includesInvalid] = useInvalid(Object.keys(Space.PUBLIC_PROPS))
  includesInvalid(rest)
  // Generate list of react styles
  const style: { width?: React.ReactText; height?: React.ReactText } = {}
  if (validWidth) style.width = validWidth
  if (validHeight) style.height = validHeight
  return <div className="Spacer" style={style} aria-hidden />
}
Space.HEIGHTS = HEIGHTS
Space.WIDTHS = WIDTHS
Space.PUBLIC_PROPS = {}
