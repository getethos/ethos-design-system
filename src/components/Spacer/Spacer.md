```jsx
import { Body } from '../Type'
;<div>
  <Body.Regular400>Some text copy</Body.Regular400>
  <Spacer.H24 />
  <Body.Regular400>Some more text copy</Body.Regular400>
  <Spacer.H24 />
</div>
```

```jsx
import { Body } from '../Type'
;<div>
  <Body.Regular400>
    Some text.
    <Spacer.W8 />
    Some more text.
  </Body.Regular400>
</div>
```
