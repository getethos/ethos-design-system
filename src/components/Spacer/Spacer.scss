/* @getethos/design-system/Spacer.scss
   ========================================================================== */

@import '../Media/Media.scss';

:root {
  --Space-80: 80px;
  --Space-72: 72px;
  --Space-64: 64px;
  --Space-56: 56px;
  --Space-48: 48px;
  --Space-40: 40px;
  --Space-32: 32px;
  --Space-24: 24px;
  --Space-16: 16px;
  --Space-8: 8px;
  --Space-4: 4px;

  --NegativeSpace-80: -80px;
  --NegativeSpace-72: -72px;
  --NegativeSpace-64: -64px;
  --NegativeSpace-56: -56px;
  --NegativeSpace-48: -48px;
  --NegativeSpace-40: -40px;
  --NegativeSpace-32: -32px;
  --NegativeSpace-24: -24px;
  --NegativeSpace-16: -16px;
  --NegativeSpace-8: -8px;
  --NegativeSpace-4: -4px;
}

@include for-phone-only {
  .spacer-height-phone-4 {
    height: var(--Space-4);
  }
  .spacer-height-phone-8 {
    height: var(--Space-8);
  }
  .spacer-height-phone-16 {
    height: var(--Space-16);
  }
  .spacer-height-phone-24 {
    height: var(--Space-24);
  }
  .spacer-height-phone-32 {
    height: var(--Space-32);
  }
  .spacer-height-phone-40 {
    height: var(--Space-40);
  }
  .spacer-height-phone-48 {
    height: var(--Space-48);
  }
  .spacer-height-phone-56 {
    height: var(--Space-56);
  }
  .spacer-height-phone-64 {
    height: var(--Space-64);
  }
  .spacer-height-phone-72 {
    height: var(--Space-72);
  }
  .spacer-height-phone-80 {
    height: var(--Space-80);
  }
  .spacer-width-phone-4 {
    width: var(--Space-4);
  }
  .spacer-width-phone-8 {
    width: var(--Space-8);
  }
  .spacer-width-phone-16 {
    width: var(--Space-16);
  }
  .spacer-width-phone-24 {
    width: var(--Space-24);
  }
  .spacer-width-phone-32 {
    width: var(--Space-32);
  }
  .spacer-width-phone-40 {
    width: var(--Space-40);
  }
  .spacer-width-phone-48 {
    width: var(--Space-48);
  }
  .spacer-width-phone-56 {
    width: var(--Space-56);
  }
  .spacer-width-phone-64 {
    width: var(--Space-64);
  }
  .spacer-width-phone-72 {
    width: var(--Space-72);
  }
  .spacer-width-phone-80 {
    width: var(--Space-80);
  }
}

@include for-tablet-only {
  .spacer-height-tablet-4 {
    height: var(--Space-4);
  }
  .spacer-height-tablet-8 {
    height: var(--Space-8);
  }
  .spacer-height-tablet-16 {
    height: var(--Space-16);
  }
  .spacer-height-tablet-24 {
    height: var(--Space-24);
  }
  .spacer-height-tablet-32 {
    height: var(--Space-32);
  }
  .spacer-height-tablet-40 {
    height: var(--Space-40);
  }
  .spacer-height-tablet-48 {
    height: var(--Space-48);
  }
  .spacer-height-tablet-56 {
    height: var(--Space-56);
  }
  .spacer-height-tablet-64 {
    height: var(--Space-64);
  }
  .spacer-height-tablet-72 {
    height: var(--Space-72);
  }
  .spacer-height-tablet-80 {
    height: var(--Space-80);
  }
  .spacer-width-tablet-4 {
    width: var(--Space-4);
  }
  .spacer-width-tablet-8 {
    width: var(--Space-8);
  }
  .spacer-width-tablet-16 {
    width: var(--Space-16);
  }
  .spacer-width-tablet-24 {
    width: var(--Space-24);
  }
  .spacer-width-tablet-32 {
    width: var(--Space-32);
  }
  .spacer-width-tablet-40 {
    width: var(--Space-40);
  }
  .spacer-width-tablet-48 {
    width: var(--Space-48);
  }
  .spacer-width-tablet-56 {
    width: var(--Space-56);
  }
  .spacer-width-tablet-64 {
    width: var(--Space-64);
  }
  .spacer-width-tablet-72 {
    width: var(--Space-72);
  }
  .spacer-width-tablet-80 {
    width: var(--Space-80);
  }
}

@include for-laptop-only {
  .spacer-height-laptop-4 {
    height: var(--Space-4);
  }
  .spacer-height-laptop-8 {
    height: var(--Space-8);
  }
  .spacer-height-laptop-16 {
    height: var(--Space-16);
  }
  .spacer-height-laptop-24 {
    height: var(--Space-24);
  }
  .spacer-height-laptop-32 {
    height: var(--Space-32);
  }
  .spacer-height-laptop-40 {
    height: var(--Space-40);
  }
  .spacer-height-laptop-48 {
    height: var(--Space-48);
  }
  .spacer-height-laptop-56 {
    height: var(--Space-56);
  }
  .spacer-height-laptop-64 {
    height: var(--Space-64);
  }
  .spacer-height-laptop-72 {
    height: var(--Space-72);
  }
  .spacer-height-laptop-80 {
    height: var(--Space-80);
  }
  .spacer-width-laptop-4 {
    width: var(--Space-4);
  }
  .spacer-width-laptop-8 {
    width: var(--Space-8);
  }
  .spacer-width-laptop-16 {
    width: var(--Space-16);
  }
  .spacer-width-laptop-24 {
    width: var(--Space-24);
  }
  .spacer-width-laptop-32 {
    width: var(--Space-32);
  }
  .spacer-width-laptop-40 {
    width: var(--Space-40);
  }
  .spacer-width-laptop-48 {
    width: var(--Space-48);
  }
  .spacer-width-laptop-56 {
    width: var(--Space-56);
  }
  .spacer-width-laptop-64 {
    width: var(--Space-64);
  }
  .spacer-width-laptop-72 {
    width: var(--Space-72);
  }
  .spacer-width-laptop-80 {
    width: var(--Space-80);
  }
}

@include for-desktop-only {
  .spacer-height-desktop-4 {
    height: var(--Space-4);
  }
  .spacer-height-desktop-8 {
    height: var(--Space-8);
  }
  .spacer-height-desktop-16 {
    height: var(--Space-16);
  }
  .spacer-height-desktop-24 {
    height: var(--Space-24);
  }
  .spacer-height-desktop-32 {
    height: var(--Space-32);
  }
  .spacer-height-desktop-40 {
    height: var(--Space-40);
  }
  .spacer-height-desktop-48 {
    height: var(--Space-48);
  }
  .spacer-height-desktop-56 {
    height: var(--Space-56);
  }
  .spacer-height-desktop-64 {
    height: var(--Space-64);
  }
  .spacer-height-desktop-72 {
    height: var(--Space-72);
  }
  .spacer-height-desktop-80 {
    height: var(--Space-80);
  }
  .spacer-width-desktop-4 {
    width: var(--Space-4);
  }
  .spacer-width-desktop-8 {
    width: var(--Space-8);
  }
  .spacer-width-desktop-16 {
    width: var(--Space-16);
  }
  .spacer-width-desktop-24 {
    width: var(--Space-24);
  }
  .spacer-width-desktop-32 {
    width: var(--Space-32);
  }
  .spacer-width-desktop-40 {
    width: var(--Space-40);
  }
  .spacer-width-desktop-48 {
    width: var(--Space-48);
  }
  .spacer-width-desktop-56 {
    width: var(--Space-56);
  }
  .spacer-width-desktop-64 {
    width: var(--Space-64);
  }
  .spacer-width-desktop-72 {
    width: var(--Space-72);
  }
  .spacer-width-desktop-80 {
    width: var(--Space-80);
  }
}
