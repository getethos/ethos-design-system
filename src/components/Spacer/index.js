export { Spacer } from './Spacer.js'
