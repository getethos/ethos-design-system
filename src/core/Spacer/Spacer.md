### Essentially deprecated

90% of uses for a spacer component in design specs is at least slightly responsive (16px at phone width, 24 at tablet, etc.)

The ideal version of Spacer is probably `<Spacer width={[16, 24, 24, 32]}/>` or something else which encourages programmers to double check the spacing at every breakpoint.

Leftover code:

```jsx
import { Body } from '../Type'
;<div>
  <Body.Regular400>Some text above a spacer</Body.Regular400>
  <Spacer.H24 />
  <Body.Regular400>Some more text</Body.Regular400>
</div>
```

```jsx
import { Body } from '../Type'
;<div>
  <Body.Regular400>
    Some text.
    <Spacer.W8 />
    Some more text.
  </Body.Regular400>
</div>
```
