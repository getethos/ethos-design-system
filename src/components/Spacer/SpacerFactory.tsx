import * as React from 'react'
import { Space } from './Space'
export function SpacerFactory(privateProps) {
  const PublicSpacerComponent = (downstreamProps) => {
    return <Space {...downstreamProps} {...privateProps} />
  }
  return PublicSpacerComponent
}
