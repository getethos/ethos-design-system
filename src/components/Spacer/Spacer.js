import { Space } from './Space'
import { SpacerFactory } from './SpacerFactory'

export const Spacer = {
  H80: SpacerFactory({ height: Space.HEIGHTS.H80 }),
  H72: SpacerFactory({ height: Space.HEIGHTS.H72 }),
  H64: SpacerFactory({ height: Space.HEIGHTS.H64 }),
  H56: SpacerFactory({ height: Space.HEIGHTS.H56 }),
  H48: SpacerFactory({ height: Space.HEIGHTS.H48 }),
  H40: SpacerFactory({ height: Space.HEIGHTS.H40 }),
  H32: SpacerFactory({ height: Space.HEIGHTS.H32 }),
  H24: SpacerFactory({ height: Space.HEIGHTS.H24 }),
  H16: SpacerFactory({ height: Space.HEIGHTS.H16 }),
  H8: SpacerFactory({ height: Space.HEIGHTS.H8 }),
  H4: SpacerFactory({ height: Space.HEIGHTS.H4 }),
  W80: SpacerFactory({ width: Space.WIDTHS.W80 }),
  W72: SpacerFactory({ width: Space.WIDTHS.W72 }),
  W64: SpacerFactory({ width: Space.WIDTHS.W64 }),
  W56: SpacerFactory({ width: Space.WIDTHS.W56 }),
  W48: SpacerFactory({ width: Space.WIDTHS.W48 }),
  W40: SpacerFactory({ width: Space.WIDTHS.W40 }),
  W32: SpacerFactory({ width: Space.WIDTHS.W32 }),
  W24: SpacerFactory({ width: Space.WIDTHS.W24 }),
  W16: SpacerFactory({ width: Space.WIDTHS.W16 }),
  W8: SpacerFactory({ width: Space.WIDTHS.W8 }),
  W4: SpacerFactory({ width: Space.WIDTHS.W4 }),
}
