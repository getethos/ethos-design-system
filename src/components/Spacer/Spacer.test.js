import React from 'react'
import { Spacer } from './Spacer.js'
import renderer from 'react-test-renderer'

describe('Spacer', () => {
  describe('matches snapshot', () => {
    test('Spacer.H80', () => {
      expect(deepSnapshot(<Spacer.H80 />)).toMatchSnapshot()
    })

    test('Spacer.H72', () => {
      expect(deepSnapshot(<Spacer.H72 />)).toMatchSnapshot()
    })

    test('Spacer.H64', () => {
      expect(deepSnapshot(<Spacer.H64 />)).toMatchSnapshot()
    })

    test('Spacer.H56', () => {
      expect(deepSnapshot(<Spacer.H56 />)).toMatchSnapshot()
    })

    test('Spacer.H48', () => {
      expect(deepSnapshot(<Spacer.H48 />)).toMatchSnapshot()
    })

    test('Spacer.H40', () => {
      expect(deepSnapshot(<Spacer.H40 />)).toMatchSnapshot()
    })

    test('Spacer.H32', () => {
      expect(deepSnapshot(<Spacer.H32 />)).toMatchSnapshot()
    })

    test('Spacer.H24', () => {
      expect(deepSnapshot(<Spacer.H24 />)).toMatchSnapshot()
    })

    test('Spacer.H16', () => {
      expect(deepSnapshot(<Spacer.H16 />)).toMatchSnapshot()
    })

    test('Spacer.H8', () => {
      expect(deepSnapshot(<Spacer.H8 />)).toMatchSnapshot()
    })

    test('Spacer.H4', () => {
      expect(deepSnapshot(<Spacer.H4 />)).toMatchSnapshot()
    })

    test('Spacer.W80', () => {
      expect(deepSnapshot(<Spacer.W80 />)).toMatchSnapshot()
    })

    test('Spacer.W72', () => {
      expect(deepSnapshot(<Spacer.W72 />)).toMatchSnapshot()
    })

    test('Spacer.W64', () => {
      expect(deepSnapshot(<Spacer.W64 />)).toMatchSnapshot()
    })

    test('Spacer.W64', () => {
      expect(deepSnapshot(<Spacer.W64 />)).toMatchSnapshot()
    })

    test('Spacer.W56', () => {
      expect(deepSnapshot(<Spacer.W56 />)).toMatchSnapshot()
    })

    test('Spacer.W48', () => {
      expect(deepSnapshot(<Spacer.W48 />)).toMatchSnapshot()
    })

    test('Spacer.W40', () => {
      expect(deepSnapshot(<Spacer.W40 />)).toMatchSnapshot()
    })

    test('Spacer.W32', () => {
      expect(deepSnapshot(<Spacer.W32 />)).toMatchSnapshot()
    })

    test('Spacer.W24', () => {
      expect(deepSnapshot(<Spacer.W24 />)).toMatchSnapshot()
    })

    test('Spacer.W16', () => {
      expect(deepSnapshot(<Spacer.W16 />)).toMatchSnapshot()
    })

    test('Spacer.W8', () => {
      expect(deepSnapshot(<Spacer.W8 />)).toMatchSnapshot()
    })

    test('Spacer.W4', () => {
      expect(deepSnapshot(<Spacer.W4 />)).toMatchSnapshot()
    })
  })

  describe('API', () => {
    test('Spacer exports properly', () => {
      expect(Spacer).toBeDefined()
      expect(Spacer.H80).toBeDefined()
      expect(Spacer.H72).toBeDefined()
      expect(Spacer.H64).toBeDefined()
      expect(Spacer.H56).toBeDefined()
      expect(Spacer.H48).toBeDefined()
      expect(Spacer.H40).toBeDefined()
      expect(Spacer.H32).toBeDefined()
      expect(Spacer.H24).toBeDefined()
      expect(Spacer.H16).toBeDefined()
      expect(Spacer.H8).toBeDefined()
      expect(Spacer.H4).toBeDefined()
      expect(Spacer.W80).toBeDefined()
      expect(Spacer.W72).toBeDefined()
      expect(Spacer.W64).toBeDefined()
      expect(Spacer.W64).toBeDefined()
      expect(Spacer.W56).toBeDefined()
      expect(Spacer.W48).toBeDefined()
      expect(Spacer.W40).toBeDefined()
      expect(Spacer.W32).toBeDefined()
      expect(Spacer.W24).toBeDefined()
      expect(Spacer.W16).toBeDefined()
      expect(Spacer.W8).toBeDefined()
      expect(Spacer.W4).toBeDefined()
    })
  })
})

export function deepSnapshot(jsx) {
  return renderer.create(jsx).toJSON()
}
