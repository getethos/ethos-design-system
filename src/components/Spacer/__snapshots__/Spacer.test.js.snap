// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer matches snapshot Spacer.H4 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 4,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H8 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 8,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H16 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 16,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H24 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 24,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H32 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 32,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H40 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 40,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H48 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 48,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H56 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 56,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H64 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 64,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H72 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 72,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.H80 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "height": 80,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W4 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 4,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W8 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 8,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W16 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 16,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W24 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 24,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W32 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 32,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W40 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 40,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W48 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 48,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W56 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 56,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W64 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 64,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W64 2`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 64,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W72 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 72,
    }
  }
/>
`;

exports[`Spacer matches snapshot Spacer.W80 1`] = `
<div
  aria-hidden={true}
  className="Spacer"
  style={
    Object {
      "width": 80,
    }
  }
/>
`;
